dosbox-staging/dosbox-staging
# Port Voodoo support with multi-threaded software backend

To test this:

- Linux / macOS / Windows MSYS builds: `meson setup -Dexperimental_voodoo_support=true build`
- Windows MSVC build: *manually* switch `C_VOODOO` define in `src/platform/visualc/config.h` - I expect the build will fail, but can't test it myself ATM.

--


![image](https://user-images.githubusercontent.com/3967/232162434-3904be7c-667a-4d1e-afa0-73cf21f7f5e0.png)
*Tomb Raider I using 3dfx patch, with viewport set to 1280x960 and `crt/geom.tweaked` shader.*

![image](https://user-images.githubusercontent.com/3967/232164500-5d237ff3-3c80-41db-b76d-b70c8da9bf15.png)
*[Screamer Rally Demo](https://www.dosgamesarchive.com/download/screamer-rally) seems to be working fine.*

--
This is another attempt at #339, this time using @kekko-dos patch with major revision by @schellingb (ec0cd9d880287440d028d9ae03c5aa2308ca384f) and published [here](https://www.vogons.org/viewtopic.php?p=1099414#p1099414). The patch size went from ~10k LOC to ~7k LOC (with lots of ifdefed-out code, but I think it might be ok for the initial merge). It's definitely easier to understand now and it's *almost-reviewable* ;).

We kept 3dfx support out of main branch for so long due to [multiple serious issues](https://github.com/dosbox-staging/dosbox-staging/issues/339#issuecomment-623049560), and this version improves the situation considerably:

1. [x] Avoid duplicated OpenGL initialization -- *No longer an issue, this version supports software backend only.*
2. [x] Provide buildsystem switch for the feature -- *Patch did it, I made some revisions to make it work correctly in Meson.* 
3. [x] Clear-out legal and licensing issues with MAME project - *Not solved yet, but not a huge blocker any more - license text embedded in the original patch was actually verbatim copy of BSD-3-Clause. We still have some work to do here though, as it turns out we're not distributing BSD license text correctly :disappointed: (AFAIK no DOSBox version does it correctly). But this should be an easy fix. (solved)*
4. [x] Window being re-created when Vodoo renders stuff -- *Solved; software renderer does not touch emulator window code at all.*
5. [x] Bad performance in software mode -- *Partly solved - for me code works smoothly in window mode and then struggles in fullscreen. This is good enough for experimental support IMHO, we can fix it later.*
6. [x] Documentation about how to properly test specific DOS games that support 3dfx. -- *I tested with Tomb Raider, but it's worth mentioning that I had to find an original 3dfx patch (and not use hacked binary distributed on GOG and Steam). It would be very helpful if someone created a wiki page and started listing links to 3dfx patches for DOS games. (update) wiki page started: https://github.com/dosbox-staging/dosbox-staging/wiki/3dfx-games*

--

@ThomasEricB can you test this branch on your SteamDeck just to confirm that #2361 is not reproducible?

@kcgen This PR catapults us to step (2) for the plan we came up with @ThomasEricB on Discord. I plan to do some small improvements in this PR still, but the amount of code is massive so let's discuss if we should drop this code all at once or cut it into smaller pieces. Let's start the review, if it turns out to be too big I will move it to draft. I don't want to fix all the problems with this code all at once - the goal is still to merge it disabled by default.
